source/common/common/assert.h

```diff
@@ -1,8 +1,54 @@
 #pragma once
 
+#include <functional>
+
 #include "common/common/logger.h"
 
 namespace Envoy {
+namespace Assert {
+
+class ActionRegistration {
+public:
+  virtual ~ActionRegistration() {}
+};
+typedef std::unique_ptr<ActionRegistration> ActionRegistrationPtr;
+
+/**
+ * Sets an action to be invoked when a debug assertion failure is detected
+ * in a release build. This action will be invoked each time an assertion
+ * failure is detected.
+ *
+ * This function is not thread-safe; concurrent calls are not allowed.
+ *
+ * This has no effect in debug builds (assertion failure aborts the process)
+ * or in release builds without ENVOY_LOG_DEBUG_ASSERT_IN_RELEASE defined (assertion
+ * tests are compiled out).
+ *
+ * @param action The action to take when an assertion fails.
+ * @return A registration object. The registration is removed when the object is destructed.
+ */
+ActionRegistrationPtr setDebugAssertionFailureRecordAction(std::function<void()> action);
+
+/**
+ * Invokes the action set by setDebugAssertionFailureRecordAction, or does nothing if
+ * no action has been set.
+ *
+ * This should only be called by ASSERT macros in this file.
+ */
+void invokeDebugAssertionFailureRecordAction_ForAssertMacroUseOnly();
```

Comment:
For my edification why is this indirection needed?

Reply:
For a little bit cleaner encapsulation, and it probably keeps the emitted code a bit smaller. But the stored value could be extern'd here, and that would work too.